versioned_hdf5/subchunk_map.py

```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
+    [((0, 2), 0), ((0, 3), 3), ((1, 0), 4)]
+    >>> _modified_chunks_in_selection(chunk_states, mappers, True)
+    [((0, 2), 0), ((0, 3), 3), ((1, 0), 4), ((1, 1), -1)]
+
+    chunk_indices        = ([0, 1, 2, 3], [1, 2])
+    whole chunks_indices = ([   1, 2   ], [1, 2])
+
+    chunk_states    selection    plus_whole=False    plus_whole=True
+    00140           pppp.        ..14.               ..14.
+    50000           pwwp.        5....               50...
+    00203           .....        .....               .....
+
+                    (p=partial,  [((0, 2), 0),       [((0, 2), 0),
+                        w=whole)  ((0, 3), 3),        ((0, 3), 3),
+                                  ((1, 0), 4)]        ((1, 0), 4),
+                                                      ((1, 1), -1)]
+    """
+    axis: Py_ssize_t
+    mapper: IndexChunkMapper
+
+    if plus_whole:
+        whole_indexers = []
+        npartial = 0
+        for mapper in mappers:
+            idx = mapper.chunks_indexer
+            widx = mapper.whole_chunks_indexer
+            whole_indexers.append(widx)
+            if isinstance(idx, slice):
+                npartial += isinstance(widx, slice) and widx != idx
+            else:
+                assert isinstance(idx, np.ndarray)
+                npartial += isinstance(widx, np.ndarray) and widx.size < idx.size
+
+        if npartial == 0:
+            raise ChunkMapIndexError()  # All chunks are wholly selected
+
+        # Add chunks that are wholly selected along all axes to the mask
+        wholes = np.zeros_like(chunk_states)
+        for axis, widx in enumerate(whole_indexers):
+            widx_nd: list = [slice(None)] * wholes.ndim
+            widx_nd[axis] = widx
+            wholes[tuple(widx_nd)] += 1
+
+    # Slice chunk_states
+    indexers = tuple(mapper.chunks_indexer for mapper in mappers)
+    indexers = _independent_ndindex(indexers)
+    states = chunk_states[indexers]
+    assert states.ndim == chunk_states.ndim == len(mappers)
+
+    mask = states != 0
+    if plus_whole:
+        wholes = wholes[indexers]
+        assert wholes.shape == mask.shape
+        mask |= wholes == len(mappers)
+
+    indices = []
+    for axis, mapper in enumerate(mappers):
+        # Get all the chunk indices involved in the selection
+        indices_i = np.asarray(mapper.chunk_indices)
+        assert indices_i.shape == (states.shape[axis],)
+        
+        # Change shape to one that matches the states
+        new_shape = tuple(-1 if i == axis else 1 for i in range(chunk_states.ndim))
+        indices_i = indices_i.reshape(new_shape)
+        indices_i = np.broadcast_to(indices_i, states.shape)
+        # Get only the indices of modified chunks
+        indices_i = indices_i[mask]
+        indices.append(indices_i.tolist())
+
+    values_idx = (states[mask] - 1).tolist()
+    return [(tuple(idx), value_idx) for *idx, value_idx in zip(*indices, values_idx)]
+
+
+def _hypercubes_between(
```

Comment:
I'm not happy with this design. It performs particularly badly in some edge cases, e.g. it's great after you call
`a[::2 * chunk_size, :] = 1`
but it's awful after
`a[:, ::2 * chunk_size] = 1`
because in the latter case you'll end up loading one chunk at a time.

It's also problematically slow when you have most of the dataset staged in memory, e.g. after `a[:] = 1`

For the sake of moving things forward, I would like to deliver like this as is and rework it later on with a smarter algorithm that can generate long and narrow rectangles.